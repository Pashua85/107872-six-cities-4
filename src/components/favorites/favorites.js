import React from 'react';
import PropsTypes from 'prop-types';
import Header from '../header/header';
import {connect} from 'react-redux';
import FavoritesPlaceCard from '../favorites-place-card/favorites-place-card';
import OffersOperation from '../../store/operations/offers-operation/offers-operation';
import {getCitiesWithFavoriteOffers} from '../../store/reducers/favorite-offers-reducer/selectors';

class Favorites extends React.PureComponent {
  constructor(props) {
    super(props);
    this.handleFavoriteClick = this.handleFavoriteClick.bind(this);
  }

  componentDidMount() {
    this.props.onComponentMount();
  }

  handleFavoriteClick(id) {
    this.props.onDeleteFromFavorite(id);
  }

  render() {
    const {citiesWithFavoriteOffers} = this.props;

    if (citiesWithFavoriteOffers.length === 0) {
      return (
        <div className="page page--favorites-empty">
          <Header />

          <main className="page__main page__main--favorites page__main--favorites-empty">
            <div className="page__favorites-container container">
              <section className="favorites favorites--empty">
                <h1 className="visually-hidden">Favorites (empty)</h1>
                <div className="favorites__status-wrapper">
                  <b className="favorites__status">Nothing yet saved.</b>
                  <p className="favorites__status-description">Save properties to narrow down search or plan yor future trips.</p>
                </div>
              </section>
            </div>
          </main>
          <footer className="footer">
            <a className="footer__logo-link" href="main.html">
              <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width="64" height="33" />
            </a>
          </footer>
        </div>
      );
    }

    return (
      <div className="page">
        <Header />

        <main className="page__main page__main--favorites">
          <div className="page__favorites-container container">
            <section className="favorites">
              <h1 className="favorites__title">Saved listing</h1>
              <ul className="favorites__list">
                {
                  citiesWithFavoriteOffers.map((c) => (
                    <li className="favorites__locations-items" key={c.city}>
                      <div className="favorites__locations locations locations--current">
                        <div className="locations__item">
                          <a className="locations__item-link" href="#">
                            <span>{c.city}</span>
                          </a>
                        </div>
                      </div>
                      <div className="favorites__places">
                        {
                          c.offers.map((of) => (
                            <FavoritesPlaceCard
                              key={of.id}
                              place={of}
                              onCardHover={() => {}}
                              onCardUnhover={() => {}}
                              styleObject={{}}
                              onCardClick={() => {}}
                            />
                          ))
                        }
                      </div>

                    </li>
                  ))
                }

              </ul>
            </section>
          </div>
        </main>
        <footer className="footer container">
          <a className="footer__logo-link" href="main.html">
            <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width="64" height="33" />
          </a>
        </footer>
      </div>
    );
  }
}

Favorites.propTypes = {
  citiesWithFavoriteOffers: PropsTypes.array.isRequired,
  onComponentMount: PropsTypes.func.isRequired,
  onDeleteFromFavorite: PropsTypes.func.isRequired
};

const mapStateToProps = (state) => ({
  citiesWithFavoriteOffers: getCitiesWithFavoriteOffers(state)
});

const mapDispatchToProps = (dispatch) => ({
  onComponentMount: () => {
    dispatch(OffersOperation.loadFavoriteOffers());
  },
  onDeleteFromFavorite: (id) => {
    dispatch(OffersOperation.deleteFromFavorite(id));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(Favorites);
export {Favorites};
