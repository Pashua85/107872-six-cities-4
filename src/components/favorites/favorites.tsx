import React, {Dispatch} from 'react';
import Header from '../header/header';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import FavoritesPlaceCard from '../favorites-place-card/favorites-place-card';
import OffersOperation from '../../store/operations/offers-operation/offers-operation';
import {getCitiesWithFavoriteOffers} from '../../store/reducers/favorite-offers-reducer/selectors';
import {ICityWithFavoriteOffers, IStore} from '../../types/types';

export interface FavoritesProps {
  citiesWithFavoriteOffers: ICityWithFavoriteOffers[],
  onComponentMount: () => void,
}


class Favorites extends React.PureComponent<FavoritesProps> {

  componentDidMount(): void {
    this.props.onComponentMount();
  }

  render(): React.ReactNode {
    const {citiesWithFavoriteOffers} = this.props;

    if (citiesWithFavoriteOffers.length === 0) {
      return (
        <div className="page page--favorites-empty">
          <Header />

          <main className="page__main page__main--favorites page__main--favorites-empty">
            <div className="page__favorites-container container">
              <section className="favorites favorites--empty">
                <h1 className="visually-hidden">Favorites (empty)</h1>
                <div className="favorites__status-wrapper">
                  <b className="favorites__status">Nothing yet saved.</b>
                  <p className="favorites__status-description">Save properties to narrow down search or plan yor future trips.</p>
                </div>
              </section>
            </div>
          </main>
          <footer className="footer">
            <a className="footer__logo-link" href="main.html">
              <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width="64" height="33" />
            </a>
          </footer>
        </div>
      );
    }

    return (
      <div className="page">
        <Header />

        <main className="page__main page__main--favorites">
          <div className="page__favorites-container container">
            <section className="favorites">
              <h1 className="favorites__title">Saved listing</h1>
              <ul className="favorites__list">
                {
                  citiesWithFavoriteOffers.map((c) => (
                    <li className="favorites__locations-items" key={c.city}>
                      <div className="favorites__locations locations locations--current">
                        <div className="locations__item">
                          <a className="locations__item-link" href="#">
                            <span>{c.city}</span>
                          </a>
                        </div>
                      </div>
                      <div className="favorites__places">
                        {
                          c.offers.map((of) => (
                            <FavoritesPlaceCard
                              key={of.id}
                              place={of}
                              onCardHover={() => {}}
                              onCardUnhover={() => {}}
                            />
                          ))
                        }
                      </div>

                    </li>
                  ))
                }

              </ul>
            </section>
          </div>
        </main>
        <footer className="footer container">
          <Link className="footer__logo-link" to="/">
            <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width="64" height="33" />
          </Link>
        </footer>
      </div>
    );
  }
}

const mapStateToProps = (state: IStore) => ({
  citiesWithFavoriteOffers: getCitiesWithFavoriteOffers(state)
});

const mapDispatchToProps = (dispatch: Dispatch<any>) => ({
  onComponentMount: () => {
    dispatch(OffersOperation.loadFavoriteOffers());
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(Favorites);
export {Favorites};
