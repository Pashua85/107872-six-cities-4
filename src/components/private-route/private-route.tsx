import React from 'react';
import {Route, Redirect, RouteProps} from 'react-router';
import {getAuthStatus} from '../../store/reducers/auth-status-reducer/selectors';
import {IStore, Status} from '../../types/types';
import {connect} from 'react-redux';
import {AUTH_STATUS} from '../../store/reducers/auth-status-reducer/auth-status-reducer';

interface PrivateRouteProps extends RouteProps {
  authStatus: Status,
  redirectPath: string
}

const PrivateRoute: React.FC<PrivateRouteProps> = (props) => (
  props.authStatus === AUTH_STATUS.AUTH ?
    <Route {...props} component={props.component} /> :
    <Redirect to={{pathname: props.redirectPath}} />
);

const mapStateToProps = (state: IStore) => ({
  authStatus: getAuthStatus(state)
});

export default connect(mapStateToProps)(PrivateRoute);
export {PrivateRoute};
