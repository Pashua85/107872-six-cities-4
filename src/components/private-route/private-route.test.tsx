import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {PrivateRoute} from './private-route';
import {Favorites} from '../favorites/favorites';

Enzyme.configure({
  adapter: new Adapter()
});

describe(`PrivateRoute`, () => {
  it(`should render correctly for authenticated user`, () => {
    const wrapper = shallow(<PrivateRoute exact path="/favorites" component={Favorites} redirectPath="/login" authStatus="AUTH" />);
    expect(wrapper).toMatchSnapshot();
  });

  it(`should render correctly for nonauthenticated guest`, () => {
    const wrapper = shallow(<PrivateRoute exact path="/favorites" component={Favorites} redirectPath="/login" authStatus="NO_AUTH" />);
    expect(wrapper).toMatchSnapshot();
  });
});