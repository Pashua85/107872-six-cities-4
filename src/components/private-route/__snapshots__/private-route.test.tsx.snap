// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute should render correctly for authenticated user 1`] = `ShallowWrapper {}`;

exports[`PrivateRoute should render correctly for nonauthenticated guest 1`] = `ShallowWrapper {}`;
